jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,6 +170,419 @@ default void clear()
+        {
+            length = Math.min(length, to.remaining());
+            to.put(bytes, offset, length);
+        }
+        finally
+        {
+            BufferUtil.flipToFlush(to, pos);
+        }
+        return length;
+    }
+
+    default void append(ByteBuffer bytes)
+    {
+        BufferUtil.append(getByteBuffer(), bytes);
+    }
+
+    default boolean append(ByteBuffer bytes, Runnable releaser)
+    {
+        append(bytes);
+        if (bytes.hasRemaining())
+            return false;
+        releaser.run();
+        return true;
+    }
+
+    default void append(RetainableByteBuffer bytes)
+    {
+        bytes.writeTo(this);
+    }
+
+    default void putTo(ByteBuffer toInfillMode)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
+        toInfillMode.put(slice);
+    }
+
+    default boolean writeTo(ByteBuffer to)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
+        BufferUtil.append(to, slice);
+        return slice.remaining() == 0;
+    }
+
+    default boolean writeTo(RetainableByteBuffer to)
+    {
+        if (remaining() == 0)
+            return true;
+        ByteBuffer slice = getByteBuffer().slice();
+        to.append(slice);
+        return slice.remaining() == 0;
+    }
+
+    default void writeTo(OutputStream out) throws IOException
+    {
+        BufferUtil.writeTo(getByteBuffer(), out);
+    }
+
+    default void writeTo(Content.Sink sink, boolean last, Callback callback)
+    {
+        sink.write(last, getByteBuffer(), callback);
+    }
+
+    static RetainableByteBuffer newAggregator(ByteBufferPool pool, boolean direct, int growBy, int maxCapacity)
+    {
+        return new Aggregator(pool, direct, growBy, maxCapacity);
+    }
+
+    class Aggregator implements RetainableByteBuffer
+    {
+        private final ByteBufferPool _pool;
+        private final boolean _direct;
+        private final int _growBy;
+        private final int _maxCapacity;
+        private RetainableByteBuffer _buffer;
+
+        private Aggregator(ByteBufferPool pool, boolean direct, int growBy, int maxCapacity)
+        {
+            if (growBy < 0)
+                throw new IllegalArgumentException("growBy must be > 0");
+            if (maxCapacity > 0 && growBy > maxCapacity)
+                throw new IllegalArgumentException("growBy must be < maxSize");
+
+            _pool = pool == null ? new ByteBufferPool.NonPooling() : pool;
+            _direct = direct;
+            _growBy = growBy;
+            _maxCapacity = maxCapacity <= 0 ? Integer.MAX_VALUE : maxCapacity;
+            _buffer = _pool.acquire(_growBy, _direct);
+        }
+
+        @Override
+        public ByteBuffer getByteBuffer()
+        {
+            return _buffer.getByteBuffer();
+        }
+
+        @Override
+        public int capacity()
+        {
+            return Math.max(_buffer.capacity(), _maxCapacity);
+        }
+
+        @Override
+        public int append(byte[] bytes, int offset, int length)
```

Comment:
I don't like this variant with `byte[]`.

Reply:
I guess we could always do the wrap externally, but it feels like a reasonable convenience method?

Reply:
Let's add it when we have a need for this method?